### PR TITLE
feat(*): Add skribbl chat

### PR DESCRIPTION
**Pull request by dhruvd-grappus**
Chat Feature

1. New web socket (api/routesws/chat.dart)
2. New url for game socket and chat socket
 ```dart
 gameRepository: GameRepository(
              uri: Uri.parse('ws://localhost:8080/ws/ws'),
            ),
            chatRepository: ChatRepository(
              uri: Uri.parse(
                'ws://localhost:8080/ws/chat',
              ),
            ),
```
See video: https://drive.google.com/file/d/147mWWRYvQ9znhFs-K6oVTvJCYqQeZOyp/view?usp=sharing

### Update
Not tested on mobile, works on web

Screenshot 2023-08-22 at 11 03 01 PM

Screenshot 2023-08-22 at 11 05 48 PM

### CODE REVIEW

**Comment by amangargamu** [comment]
Bhadiya

File: api/lib/chat/chat_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import 'package:api/chat/chat_model.dart';
```
Comment (noorbhatia10) [inline api/lib/chat/chat_bloc.dart]: I think we should keep this in session bloc, as chats will exist as long as a session exists. Making a separate connection doesn't make sense. Thoughts?

File: api/routes/ws/chat.dart [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            jsonEncode({'status': 'error', 'message': e.toString()}),
+          );
+        }
+      },
+      onDone: () {
+        chatCubit.unsubscribe(channel);
+      },
+    );
+  });
+  return handler(context);
```
Comment (noorbhatia10) [inline api/routes/ws/chat.dart]: We can do all this in our original ws connection, we just have to handle type of message added in the channel sink. Eg:  __add_player__, __update_offsets__, __send_message__

File: api/lib/chat/chat_bloc.dart (hunk 1 above)
Comment (dhruvd-grappus) [reply api/lib/chat/chat_bloc.dart]: Websocket for drawing and chat is different, separation of concerns.

File: api/routes/ws/chat.dart (hunk 2 above)
Comment (dhruvd-grappus) [reply api/routes/ws/chat.dart]: @gauravkatvaa and I were going to do this first
But wouldnt we risk performance by using 1 socket for everything?

File: api/lib/chat/chat_bloc.dart (hunk 1 above)
Comment (noorbhatia10) [reply api/lib/chat/chat_bloc.dart]: no the connection is unique, we don't establish separate connection for them

Comment (noorbhatia10) [reply api/lib/chat/chat_bloc.dart]: on FE repositories can be different but on BE chat is a part of session

Comment (dhruvd-grappus) [reply api/lib/chat/chat_bloc.dart]: makes sense, will change.

File: api/routes/ws/chat.dart (hunk 2 above)
Comment (noorbhatia10) [reply api/routes/ws/chat.dart]: Performance won't be affected and multiple ws connections can cause synchronization problems. It's better to use a single connection throughout the game. We can differentiate events like I mentioned above

Comment (mokshgrappus007) [reply api/routes/ws/chat.dart]: right we can create an enum and on that basis we can pass the data to the stream

**Comment by noorbhatia10** [comment]
Great work guys!

**Review by noorbhatia10 — APPROVED** [review]

**Pull request merged**